In [1]:
%matplotlib inline

In [2]:
import torch
import os
from torch.utils.data import Dataset
from torchvision import datasets, transforms
from torchvision.transforms import ToTensor
from torch import nn
from torch.utils.data import DataLoader
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import finalprojectneuralnetworks as fpnn
import csv
import time

In [3]:
# import random


# with open("Dummy_Validate.csv", "w", newline="") as csvfile:
#     spamwriter = csv.writer(
#         csvfile, delimiter=",", quotechar='"', quoting=csv.QUOTE_MINIMAL
#     )
#     col_labels = [
#         "AF3",
#         "F7",
#         "F3",
#         "FC5",
#         "T7",
#         "P7",
#         "O1",
#         "O2",
#         "P8",
#         "T8",
#         "FC6",
#         "F4",
#         "F8",
#         "AF4",
#         "eyeDetection",
#     ]
#     spamwriter.writerow(col_labels)
#     list_a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 0]
#     list_b = [21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 1]
#     for i in range(500):
#         random_number = random.randint(1, 10)
#         if random_number > 5:
#             spamwriter.writerow(list_a)
#         else:
#             spamwriter.writerow(list_b)

In [4]:
# root_directory = 'D:\College\CS_545\Final_Project'
# fp_csv_file = 'EEG_Eye_State.csv'
# train_fraction = 0.8
# validate_fraction = 0.1
# X = pd.read_csv(os.path.join(root_directory,fp_csv_file), delimiter=',', usecols=range(15)).dropna(axis=0).to_numpy()
# n_samples = X.shape[0]
# rows = np.arange(n_samples)
# np.random.shuffle(rows)

# n_train = round(n_samples * train_fraction)
# n_validate = round(n_samples * validate_fraction)

# col_labels = ["AF3","F7","F3","FC5","T7","P7","O1","O2","P8","T8","FC6","F4","F8","AF4","eyeDetection"]
# Xtrain = X[rows[:n_train], :]
# Xvalidate = X[rows[n_train:n_train + n_validate], :]
# Xtest = X[rows[n_train + n_validate:], :]
# Xtrain = pd.DataFrame(Xtrain, columns = col_labels)
# Xvalidate = pd.DataFrame(Xvalidate, columns = col_labels)
# Xtest = pd.DataFrame(Xtest, columns = col_labels)
# Xtrain = Xtrain.astype({"eyeDetection": int})
# Xvalidate = Xvalidate.astype({"eyeDetection": int})
# Xtest = Xtest.astype({"eyeDetection": int})
# print(Xtrain.iloc[0])
# print( type(X))
# Xtrain.to_csv("EEG_Eye_State_Train.csv", encoding='utf-8', index_label=False, index=False, quoting=csv.QUOTE_NONNUMERIC )
# Xvalidate.to_csv("EEG_Eye_State_Validate.csv", encoding='utf-8', index_label=False, index=False, quoting=csv.QUOTE_NONNUMERIC )
# Xtest.to_csv("EEG_Eye_State_Test.csv", encoding='utf-8', index_label=False, index=False, quoting=csv.QUOTE_NONNUMERIC )

In [5]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"
fp_csv_file = "EEG_Eye_State_Train.csv"
# fp_csv_file = "Dummy_Train.csv"
eeg_data = pd.read_csv(
    os.path.join(root_directory, fp_csv_file), delimiter=",", usecols=range(15)
).dropna(axis=0)

print(eeg_data.iloc[0, :])
open_or_closed = eeg_data.iloc[0, 14]
channel_data = eeg_data.iloc[0, 0:14]
channel_data = np.array([channel_data])
channel_data = channel_data.astype("float").reshape(-1, 14)
print(open_or_closed)
print(channel_data)
eeg_data.shape

AF3             4309.74
F7              4052.82
F3              4285.13
FC5             4155.38
T7              4369.23
P7              4646.67
O1              4100.00
O2              4649.74
P8              4221.54
T8              4230.26
FC6             4206.15
F4              4292.82
F8              4598.97
AF4             4362.56
eyeDetection       0.00
Name: 0, dtype: float64
0
[[4309.74 4052.82 4285.13 4155.38 4369.23 4646.67 4100.   4649.74 4221.54
  4230.26 4206.15 4292.82 4598.97 4362.56]]


(11984, 15)

In [6]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"

train_file = "EEG_Eye_State_Train.csv"
validate_file = "EEG_Eye_State_Validate.csv"
test_file = "EEG_Eye_State_Test.csv"

# train_file = "Dummy_Train.csv"
# validate_file = "Dummy_Validate.csv"
# test_file = "Dummy_Test.csv"

train_dataset = fpnn.FinalProjectEEGDataset(train_file, root_directory)
validate_dataset = fpnn.FinalProjectEEGDataset(validate_file, root_directory)
test_dataset = fpnn.FinalProjectEEGDataset(test_file, root_directory)
print(train_dataset.__len__())
train_dataset.__getitem__(0)
train_dataloader = DataLoader(train_dataset, batch_size=64)
validate_dataloader = DataLoader(validate_dataset, batch_size=64)
test_dataloader = DataLoader(test_dataset, batch_size=64)

AF3             float64
F7              float64
F3              float64
FC5             float64
T7              float64
P7              float64
O1              float64
O2              float64
P8              float64
T8              float64
FC6             float64
F4              float64
F8              float64
AF4             float64
eyeDetection      int64
dtype: object
AF3             float64
F7              float64
F3              float64
FC5             float64
T7              float64
P7              float64
O1              float64
O2              float64
P8              float64
T8              float64
FC6             float64
F4              float64
F8              float64
AF4             float64
eyeDetection      int64
dtype: object
AF3             float64
F7              float64
F3              float64
FC5             float64
T7              float64
P7              float64
O1              float64
O2              float64
P8              float64
T8              float64
FC6         

In [7]:
device = "cuda" if torch.cuda.is_available() else "cpu"
print("Using {} device".format(device))

Using cpu device


In [8]:
model = fpnn.NeuralNetwork(28 * 28, [512, 512], 10).to(device)
print(model)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=784, out_features=512, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=512, out_features=512, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=512, out_features=10, bias=True)
  )
)


In [9]:
X = torch.rand(1, 28, 28, device=device)
logits = model(X)
print(logits)
pred_probab = nn.Softmax(dim=1)(logits)
y_pred = pred_probab.argmax(1)
print(f"Predicted class: {y_pred}")

tensor([[ 0.0452, -0.0094, -0.0277,  0.0657,  0.0079,  0.0026,  0.0191,  0.0068,
          0.0008, -0.0493]], grad_fn=<AddmmBackward>)
Predicted class: tensor([3])


In [10]:
def train(dataloader, model, loss_fn, optimizer):
    size = len(dataloader.dataset)
    model.train()
    # first = True
    for batch, (X, T) in enumerate(dataloader):
        X, T = X.to(device), T.to(device)

        # Compute prediction error
        Y = model(X.float())
        # if first:
        #     first = False
        #     print('Y.shape', Y.shape)
        #     print('T.shape', T.shape)
        #     print('T: ', T[0])
        loss = loss_fn(Y, T.long())

        # Backpropagation
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()


#         if batch % 50 == 0:
#             loss, current = loss.item(), batch * len(X)
# print(f"loss: {loss:>7f}  [{current:>5d}/{size:>5d}]")

In [11]:
def test(dataloader, model, loss_fn):
    size = len(dataloader.dataset)
    num_batches = len(dataloader)
    model.eval()
    test_loss, correct = 0, 0
    with torch.no_grad():
        for X, T in dataloader:
            X, T = X.to(device), T.to(device)
            pred = model(X.float())
            test_loss += loss_fn(pred, T.long()).item()
            correct += (pred.argmax(1) == T).type(torch.float).sum().item()
    test_loss /= num_batches
    correct /= size
    print(
        f"Test Error: \n Accuracy: {(100*correct):>0.1f}%, Avg loss: {test_loss:>8f} \n"
    )

In [12]:
def train_for_epochs(
    epochs, train_dataloader, _validate_dataloader, model, loss_fn, optimizer
):
    start = time.time()
    for t in range(epochs):
        train(train_dataloader, model, loss_fn, optimizer)
        if t % 50 == 0:
            print(f"Epoch {t}\n-------------------------------")
            test(_validate_dataloader, model, loss_fn)
    end = time.time()
    print("Done: ", end - start, " seconds")
    test(_validate_dataloader, model, loss_fn)

In [13]:
def count_ones_and_zeros(file):
    data = pd.read_csv(
        os.path.join(file),
        delimiter=",",
    ).dropna(axis=0)
    zeros = 0
    ones = 1
    for index, row in data.iterrows():
        if row[len(data.columns) - 1] == 0:
            zeros += 1
        else:
            ones += 1
    print("Zeros: ", zeros)
    print("Ones: ", ones)

In [14]:
# fashion_training_data = datasets.FashionMNIST(
#     root="data",
#     train=True,
#     download=True,
#     transform=ToTensor(),
# )
# fashion_train_dataloader = DataLoader(fashion_training_data, batch_size=64)
# n_samples, n_inputs = train_dataset.__len__(), 28*28
# n_outputs = 1
# n_hiddens = [10, 10]
# model = fpnn.NeuralNetwork(28*28, n_hiddens, 10).to(device)
# print(model)

# learning_rate = 0.5 / (n_samples * n_outputs)  ## Larger learning rate

# optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
# loss_fn = torch.nn.CrossEntropyLoss()

# train(fashion_train_dataloader, model, loss_fn, optimizer)

In [15]:
for X, y in train_dataloader:
    print("Shape of X [N, C, H, W]: ", X.shape)
    print("Shape of y: ", y.shape, y.dtype)
    break

Shape of X [N, C, H, W]:  torch.Size([64, 1, 14])
Shape of y:  torch.Size([64]) torch.float64


I started out with a dummy run using dummy data that had a trivial pattern which the neural network should easily be able to learn. I did this to simply test that the network was correctly setup.

In [15]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"

# train_file = "EEG_Eye_State_Train.csv"
# validate_file = "EEG_Eye_State_Validate.csv"
# test_file = "EEG_Eye_State_Test.csv"

train_file = "Dummy_Train.csv"
validate_file = "Dummy_Validate.csv"
test_file = "Dummy_Test.csv"

train_dataset = fpnn.FinalProjectEEGDataset(train_file, root_directory)
validate_dataset = fpnn.FinalProjectEEGDataset(validate_file, root_directory)
test_dataset = fpnn.FinalProjectEEGDataset(test_file, root_directory)
train_dataloader = DataLoader(train_dataset, batch_size=64)
validate_dataloader = DataLoader(validate_dataset, batch_size=64)
test_dataloader = DataLoader(test_dataset, batch_size=64)

AF3             int64
F7              int64
F3              int64
FC5             int64
T7              int64
P7              int64
O1              int64
O2              int64
P8              int64
T8              int64
FC6             int64
F4              int64
F8              int64
AF4             int64
eyeDetection    int64
dtype: object
AF3             int64
F7              int64
F3              int64
FC5             int64
T7              int64
P7              int64
O1              int64
O2              int64
P8              int64
T8              int64
FC6             int64
F4              int64
F8              int64
AF4             int64
eyeDetection    int64
dtype: object
AF3             int64
F7              int64
F3              int64
FC5             int64
T7              int64
P7              int64
O1              int64
O2              int64
P8              int64
T8              int64
FC6             int64
F4              int64
F8              int64
AF4             int64
eyeD

In [16]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [5, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "sigmoid").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 10
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=5, bias=True)
    (Sigmoid0): Sigmoid()
    (Linear1): Linear(in_features=5, out_features=5, bias=True)
    (Sigmoid1): Sigmoid()
    (Linear2): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 50.0%, Avg loss: 0.695648 

Done:  1.8184118270874023  seconds
Test Error: 
 Accuracy: 100.0%, Avg loss: 0.595044 



In [17]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 100.0%, Avg loss: 0.534908 



Unsurprisingly in this dummy run the model learned to predict with 100% accuracy.

I then moved on to the actual data to see what kind of accuracy I could get from a classical classification neural network model. I tried several structures and parameters.

In [78]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"

train_file = "EEG_Eye_State_Train.csv"
validate_file = "EEG_Eye_State_Validate.csv"
test_file = "EEG_Eye_State_Test.csv"

# train_file = "Dummy_Train.csv"
# validate_file = "Dummy_Validate.csv"
# test_file = "Dummy_Test.csv"

train_dataset = fpnn.FinalProjectEEGDataset(train_file, root_directory)
validate_dataset = fpnn.FinalProjectEEGDataset(validate_file, root_directory)
test_dataset = fpnn.FinalProjectEEGDataset(test_file, root_directory)
train_dataloader = DataLoader(train_dataset, batch_size=64)
validate_dataloader = DataLoader(validate_dataset, batch_size=64)
test_dataloader = DataLoader(test_dataset, batch_size=64)

AF3             float64
F7              float64
F3              float64
FC5             float64
T7              float64
P7              float64
O1              float64
O2              float64
P8              float64
T8              float64
FC6             float64
F4              float64
F8              float64
AF4             float64
eyeDetection      int64
dtype: object
AF3             float64
F7              float64
F3              float64
FC5             float64
T7              float64
P7              float64
O1              float64
O2              float64
P8              float64
T8              float64
FC6             float64
F4              float64
F8              float64
AF4             float64
eyeDetection      int64
dtype: object
AF3             float64
F7              float64
F3              float64
FC5             float64
T7              float64
P7              float64
O1              float64
O2              float64
P8              float64
T8              float64
FC6         

In [79]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [5, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 10
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=5, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=5, out_features=5, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.687582 

Done:  24.560352325439453  seconds
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685850 



In [20]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [10, 10]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 501
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

i == 0
NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=10, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=10, out_features=10, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=10, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 46.4%, Avg loss: 10.459651 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 3.271880 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 1.862043 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685989 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685970 

Epoch 500
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685949 

Done:  1008.1690013408661  seconds


In [21]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [30, 20, 15, 10]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 501
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

i == 0
NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=30, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=30, out_features=20, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=20, out_features=15, bias=True)
    (ReLU2): ReLU()
    (Linear3): Linear(in_features=15, out_features=10, bias=True)
    (ReLU3): ReLU()
    (Linear4): Linear(in_features=10, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 1.173773 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685931 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685854 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685854 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.685854 

Epoch 500
----

In [22]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [500, 500]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 500
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

i == 0
NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=500, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=500, out_features=500, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=500, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 6.962534 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 1.713657 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 1.713657 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 1.713657 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 1.713657 

Done:  1214.9345533847809  seconds


In [23]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [500, 500]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.01 / (n_samples * n_outputs)

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 500
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

i == 0
NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=500, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=500, out_features=500, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=500, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 288.282979 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 54.4%, Avg loss: 0.785713 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 55.3%, Avg loss: 0.756815 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 56.1%, Avg loss: 0.722233 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 56.7%, Avg loss: 0.705566 

Done:  1155.8505518436432  seconds


In [24]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [500, 500]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.01 / (n_samples * n_outputs)

optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 500
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

i == 0
NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=500, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=500, out_features=500, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=500, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss:      nan 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss:      nan 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss:      nan 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss:      nan 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss:      nan 

Done:  1099.8602554798126  seconds


In [25]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [500, 500]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.01 / (n_samples * n_outputs)

optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 500
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

i == 0
NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_relu_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=500, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=500, out_features=500, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=500, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 77656832.685102 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 77656832.683104 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 77656832.681246 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 77656832.679520 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 77656832.677915 

Done:  1064.0559866428375  seconds


In [21]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [500, 500]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "tanh").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.01 / (n_samples * n_outputs)

optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 500
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=500, bias=True)
    (Tanh0): Tanh()
    (Linear1): Linear(in_features=500, out_features=500, bias=True)
    (Tanh1): Tanh()
    (Linear2): Linear(in_features=500, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 0.755901 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.686656 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 0.686432 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 0.686410 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 55.9%, Avg loss: 0.686408 

Done:  1249.532361984253  seconds


In [30]:
n_samples, n_inputs = train_dataset.__len__(), 14
n_outputs = 2
n_hiddens = [500, 500]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "sigmoid").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.01 / (n_samples * n_outputs)

optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 500
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=14, out_features=500, bias=True)
    (Sigmoid0): Sigmoid()
    (Linear1): Linear(in_features=500, out_features=500, bias=True)
    (Sigmoid1): Sigmoid()
    (Linear2): Linear(in_features=500, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 50.0%, Avg loss: 0.670922 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 100.0%, Avg loss: 0.006674 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 100.0%, Avg loss: 0.002450 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 100.0%, Avg loss: 0.001424 

Epoch 400
-------------------------------
Test Error: 
 Accuracy: 100.0%, Avg loss: 0.000983 

Done:  94.19774389266968  seconds


Unfortunately this approach didn't seem to produce very good results. The models were barely better than chance, with he best of them achieving a final accuracy of 56%. They also didn't seem to learn much after the initial few epochs. I decided to try a different approach going forward. 

## Including Mean and Standard Deviation Approach Experimentation ##

I started by removing any glaring outliers from the data. I thought this might make it a little easier for the neural network to learn.

In [18]:
# root_directory = "D:\College\CS_545\Final_Project"
# fp_csv_file = "EEG_Eye_State.csv"
# data = pd.read_csv(
#     os.path.join(root_directory, fp_csv_file), delimiter=",", usecols=range(15)
# ).dropna(axis=0)

# # data = pd.read_csv(os.path.join(root_directory, fp_csv_file), header=None)
# # retrieve data as numpy array
# values = data.values


# print(values.shape)
# print(values[0:5])
# # step over each EEG column
# for i in range(values.shape[1] - 1):
#     # calculate column mean and standard deviation
#     data_mean, data_std = np.mean(values[:, i]), np.std(values[:, i])
#     # define outlier bounds
#     cut_off = data_std * 4
#     lower, upper = data_mean - cut_off, data_mean + cut_off
#     # remove too small
#     too_small = [j for j in range(values.shape[0]) if values[j, i] < lower]
#     values = np.delete(values, too_small, 0)
#     print(">deleted %d rows" % len(too_small))
#     # remove too large
#     too_large = [j for j in range(values.shape[0]) if values[j, i] > upper]
#     values = np.delete(values, too_large, 0)
#     print(">deleted %d rows" % len(too_large))
# # save the results to a new file
# col_labels = [
#     "AF3",
#     "F7",
#     "F3",
#     "FC5",
#     "T7",
#     "P7",
#     "O1",
#     "O2",
#     "P8",
#     "T8",
#     "FC6",
#     "F4",
#     "F8",
#     "AF4",
#     "eyeDetection",
# ]
# values = pd.DataFrame(values, columns=data.columns)
# values.to_csv(
#     "EEG_Eye_State_No_Outliers.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

After that I decided to try and investigate the data at a more meta level. I calculated the mean, standard deviation, maximum value, and minimum value for the eye-open state data and eye-closed state data. I also calculated the mean and standard deviation for the data as a whole.

In [19]:
def get_metadata(file_directory):
    data = pd.read_csv(file_directory, delimiter=",", usecols=range(15)).dropna(axis=0)

    metadata_labels = [
        "Channel",
        "Mean",
        "Std",
        "Max",
        "Min",
        "Eye-Open Mean",
        "Eye-Open Std",
        "Eye-Open Max",
        "Eye-Open Min",
        "Eye-Closed Mean",
        "Eye-Closed Std",
        "Eye-Closed Max",
        "Eye-Closed Min",
    ]

    metadata = [
        ["AF3"],
        ["F7"],
        ["F3"],
        ["FC5"],
        ["T7"],
        ["P7"],
        ["O1"],
        ["O2"],
        ["P8"],
        ["T8"],
        ["FC6"],
        ["F4"],
        ["F8"],
        ["AF4"],
    ]
    values = data.values
    print(values.shape)
    # step over each EEG column
    for i in range(values.shape[1] - 1):
        # calculate column mean and standard deviation
        data_mean, data_std, data_max, data_min = (
            np.mean(values[:, i]),
            np.std(values[:, i]),
            np.max(values[:, i]),
            np.min(values[:, i]),
        )
        metadata[i].append(data_mean)
        metadata[i].append(data_std)
        metadata[i].append(data_max)
        metadata[i].append(data_min)

    zeros = []
    ones = []
    for sample in values:
        if sample[14] == 0:
            zeros.append(sample)
        else:
            ones.append(sample)
    zeros = pd.DataFrame(zeros)
    ones = pd.DataFrame(ones)

    zeros.columns = data.columns
    ones.columns = data.columns
    print(zeros.shape)
    zeros_values = zeros.values
    for i in range(zeros_values.shape[1] - 1):
        # calculate column mean and standard deviation
        data_mean, data_std, data_max, data_min = (
            np.mean(zeros_values[:, i]),
            np.std(zeros_values[:, i]),
            np.max(zeros_values[:, i]),
            np.min(zeros_values[:, i]),
        )
        metadata[i].append(data_mean)
        metadata[i].append(data_std)
        metadata[i].append(data_max)
        metadata[i].append(data_min)

    print(ones.shape)
    ones_values = ones.values
    for i in range(ones_values.shape[1] - 1):
        # calculate column mean and standard deviation
        data_mean, data_std, data_max, data_min = (
            np.mean(ones_values[:, i]),
            np.std(ones_values[:, i]),
            np.max(ones_values[:, i]),
            np.min(ones_values[:, i]),
        )
        metadata[i].append(data_mean)
        metadata[i].append(data_std)
        metadata[i].append(data_max)
        metadata[i].append(data_min)

    metadata = pd.DataFrame(metadata)
    metadata.columns = metadata_labels
    return metadata

In [20]:
root_directory = "D:\College\CS_545\Final_Project"
fp_csv_file = "EEG_Eye_State_No_Outliers.csv"
metadata = get_metadata(os.path.join(root_directory, fp_csv_file))
metadata

(14304, 15)
(7855, 15)
(6449, 15)


Channel         Mean        Std      Max      Min  Eye-Open Mean  \
0      AF3  4298.144530  32.383173  4466.15  4198.97    4294.614386   
1       F7  4007.199968  27.222342  4154.36  3913.33    4009.736631   
2       F3  4261.656384  16.219297  4349.23  4197.44    4260.402014   
3      FC5  4120.176654  16.904783  4191.79  4067.18    4120.948880   
4       T7  4339.673482  11.994233  4397.95  4308.72    4339.912341   
5       P7  4618.165981  12.923532  4672.31  4566.15    4619.437780   
6       O1  4070.977384  17.707910  4138.97  4026.15    4070.531945   
7       O2  4614.060973  14.398261  4672.82  4567.18    4613.460053   
8       P8  4199.641958  13.807744  4255.90  4147.69    4198.939306   
9       T8  4229.291521  15.005626  4288.21  4170.26    4227.845602   
10     FC6  4199.729761  18.338192  4271.28  4125.13    4197.989321   
11      F4  4276.866449  15.089147  4332.31  4216.41    4274.783659   
12      F8  4603.123432  25.535726  4716.41  4490.77    4599.781686   
13     AF4  4358.072562  31.876578  4491.28  4236.41    4353.527338   

    Eye-Open Std  Eye-Open Max  Eye-Open Min  Eye-Closed Mean  Eye-Closed Std  \
0      33.537700       4466.15       4206.67      4302.444309       30.369437   
1      28.256648       4154.36       3924.10      4004.110265       25.569113   
2      16.607684       4349.23       4197.44      4263.184229       15.597616   
3      16.536333       4187.18       4073.33      4119.236069       17.296483   
4      11.199247       4395.90       4308.72      4339.382548       12.890536   
5      13.217759       4671.28       4566.15      4616.616906       12.380614   
6      14.417633       4124.10       4027.18      4071.519936       21.018584   
7      13.655753       4672.82       4567.18      4614.792904       15.221895   
8      12.424151       4255.90       4152.31      4200.497801       15.281592   
9      14.217735       4287.69       4170.26      4231.052678       15.734208   
10     17.985427       4264.10       4125.13      4201.849648       18.539501   
11     15.042807       4329.74       4216.41      4279.403325       14.753427   
12     23.835106       4711.79       4490.77      4607.193740       26.910407   
13     32.223892       4491.28       4236.41      4363.608728       30.548114   

    Eye-Closed Max  Eye-Closed Min  
0          4441.03         4198.97  
1          4138.97         3913.33  
2          4333.85         4212.31  
3          4191.79         4067.18  
4          4397.95         4309.74  
5          4672.31         4581.03  
6          4138.97         4026.15  
7          4672.82         4567.69  
8          4255.38         4147.69  
9          4288.21         4174.36  
10         4271.28         4130.77  
11         4332.31         4225.64  
12         4716.41         4510.26  
13         4487.69         4246.15

The differences between various channel metadata values in the eye-open versus eye-closed date tended to be very small. They did seem to be present though. Given this I thought it would be interesting to add these values as inputs to the neural network. I thought adding these as a feature might give the neural network some information that would make it easier to identify trends between the states.

I started out by just adding the mean and standard deviation for the data as a whole as one of the data points for each sample. To avoid passing information from validate or test into train I first split up the data into train, validate, and test sets. I decided to preserve the temporal ordering this time when splitting up the data. For each of these new datasets I then calculated the overall mean and standard deviation for each channel and added those as column data in each dataset.

In [19]:
metadata.iloc[0].values

array(['AF3', 4298.144529502238, 32.383172751002164, 4466.15, 4198.97,
       4294.6143857415655, 33.53769989310272, 4466.15, 4206.67,
       4302.444309195224, 30.369436627795682, 4441.03, 4198.97],
      dtype=object)

In [25]:
# root_directory = "D:\College\CS_545\Final_Project"
# fp_csv_file = "EEG_Eye_State_No_Outliers.csv"
# data = pd.read_csv(
#     os.path.join(root_directory, fp_csv_file), delimiter=",", usecols=range(15)
# ).dropna(axis=0)

# col_labels = []
# print(data.shape)
# for i in range(metadata.shape[0]):
#     col_labels.append(metadata.iloc[i].values[0])
#     col_labels.append(metadata.iloc[i].values[0] + " " + metadata.columns[1])
#     col_labels.append(metadata.iloc[i].values[0] + " " + metadata.columns[2])
#     col_labels.append(metadata.iloc[i].values[0] + " " + metadata.columns[3])
#     col_labels.append(metadata.iloc[i].values[0] + " " + metadata.columns[4])

# col_labels.append("eyeDetection")
# print(col_labels)

# new_data = []
# # num = 0
# for index, row in data.iterrows():
#     # if num == 100:
#     #     break
#     # num += num
#     temp_list = []
#     for i in range(len(row)):
#         temp_list.append(row[i])
#         if i < len(row) - 1:
#             temp_list.append(metadata.iloc[i].values[1])
#             temp_list.append(metadata.iloc[i].values[2])
#             temp_list.append(metadata.iloc[i].values[3])
#             temp_list.append(metadata.iloc[i].values[4])
#     new_data.append(temp_list)
# print(len(new_data))

(14304, 15)
['AF3', 'AF3 Mean', 'AF3 Std', 'AF3 Max', 'AF3 Min', 'F7', 'F7 Mean', 'F7 Std', 'F7 Max', 'F7 Min', 'F3', 'F3 Mean', 'F3 Std', 'F3 Max', 'F3 Min', 'FC5', 'FC5 Mean', 'FC5 Std', 'FC5 Max', 'FC5 Min', 'T7', 'T7 Mean', 'T7 Std', 'T7 Max', 'T7 Min', 'P7', 'P7 Mean', 'P7 Std', 'P7 Max', 'P7 Min', 'O1', 'O1 Mean', 'O1 Std', 'O1 Max', 'O1 Min', 'O2', 'O2 Mean', 'O2 Std', 'O2 Max', 'O2 Min', 'P8', 'P8 Mean', 'P8 Std', 'P8 Max', 'P8 Min', 'T8', 'T8 Mean', 'T8 Std', 'T8 Max', 'T8 Min', 'FC6', 'FC6 Mean', 'FC6 Std', 'FC6 Max', 'FC6 Min', 'F4', 'F4 Mean', 'F4 Std', 'F4 Max', 'F4 Min', 'F8', 'F8 Mean', 'F8 Std', 'F8 Max', 'F8 Min', 'AF4', 'AF4 Mean', 'AF4 Std', 'AF4 Max', 'AF4 Min', 'eyeDetection']
14304


In [26]:
# print(data.iloc[0])
# print(new_data[0])
# print(len(new_data[0]))
# new_data_df = pd.DataFrame(new_data, columns=col_labels)
# new_data_df = new_data_df.round(2)
# new_data_df = new_data_df.astype({"eyeDetection": int})
# new_data_df.to_csv(
#     "EEG_Eye_State_Mean_Std.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

AF3             4329.23
F7              4009.23
F3              4289.23
FC5             4148.21
T7              4350.26
P7              4586.15
O1              4096.92
O2              4641.03
P8              4222.05
T8              4238.46
FC6             4211.28
F4              4280.51
F8              4635.90
AF4             4393.85
eyeDetection       0.00
Name: 0, dtype: float64
[4329.23, 4298.144529502238, 32.383172751002164, 4466.15, 4198.97, 4009.23, 4007.1999678411635, 27.222342363737564, 4154.36, 3913.33, 4289.23, 4261.656383529083, 16.2192974906191, 4349.23, 4197.44, 4148.21, 4120.176654082774, 16.90478308219174, 4191.79, 4067.18, 4350.26, 4339.673482242729, 11.99423301793753, 4397.95, 4308.72, 4586.15, 4618.165980844519, 12.923532182692702, 4672.31, 4566.15, 4096.92, 4070.977383948546, 17.707909625094178, 4138.97, 4026.15, 4641.03, 4614.060973154362, 14.39826115365047, 4672.82, 4567.18, 4222.05, 4199.641958193512, 13.807743808585931, 4255.9, 4147.69, 4238.46, 4229.291521252797

In [29]:
# root_directory = "D:\College\CS_545\Final_Project"
# fp_csv_file = "EEG_Eye_State_Mean_Std.csv"
# data = pd.read_csv(
#     os.path.join(root_directory, fp_csv_file), delimiter=",", usecols=range(71)
# ).dropna(axis=0)

# train_fraction = 0.8
# validate_fraction = 0.1
# X = (
#     pd.read_csv(
#         os.path.join(root_directory, fp_csv_file), delimiter=",", usecols=range(71)
#     )
#     .dropna(axis=0)
#     .to_numpy()
# )
# n_samples = X.shape[0]

# n_train = round(n_samples * train_fraction)
# n_validate = round(n_samples * validate_fraction)

# Xtrain = X[:n_train]
# Xvalidate = X[n_train : n_train + n_validate]
# Xtest = X[n_train + n_validate :]
# Xtrain = pd.DataFrame(Xtrain, columns=col_labels)
# Xvalidate = pd.DataFrame(Xvalidate, columns=col_labels)
# Xtest = pd.DataFrame(Xtest, columns=col_labels)
# Xtrain = Xtrain.astype({"eyeDetection": int})
# Xvalidate = Xvalidate.astype({"eyeDetection": int})
# Xtest = Xtest.astype({"eyeDetection": int})
# print(Xtrain.iloc[0])
# print(type(X))
# Xtrain.to_csv(
#     "EEG_Eye_State_Mean_Std_Train.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )
# Xvalidate.to_csv(
#     "EEG_Eye_State_Mean_Std_Validate.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )
# Xtest.to_csv(
#     "EEG_Eye_State_Mean_Std_Test.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

AF3             4329.23
AF3 Mean        4298.14
AF3 Std           32.38
AF3 Max         4466.15
AF3 Min         4198.97
                 ...   
AF4 Mean        4358.07
AF4 Std           31.88
AF4 Max         4491.28
AF4 Min         4236.41
eyeDetection       0.00
Name: 0, Length: 71, dtype: float64
<class 'numpy.ndarray'>


In [80]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"

train_file = "EEG_Eye_State_Mean_Std_Train.csv"
validate_file = "EEG_Eye_State_Mean_Std_Validate.csv"
test_file = "EEG_Eye_State_Mean_Std_Test.csv"

# train_file = "Dummy_Train.csv"
# validate_file = "Dummy_Validate.csv"
# test_file = "Dummy_Test.csv"

train_dataset = fpnn.FinalProjectEEGDataset(train_file, root_directory)
validate_dataset = fpnn.FinalProjectEEGDataset(validate_file, root_directory)
test_dataset = fpnn.FinalProjectEEGDataset(test_file, root_directory)
train_dataloader = DataLoader(train_dataset, batch_size=64)
validate_dataloader = DataLoader(validate_dataset, batch_size=64)
test_dataloader = DataLoader(test_dataset, batch_size=64)

AF3             float64
AF3 Mean        float64
AF3 Std         float64
AF3 Max         float64
AF3 Min         float64
                 ...   
AF4 Mean        float64
AF4 Std         float64
AF4 Max         float64
AF4 Min         float64
eyeDetection      int64
Length: 71, dtype: object
AF3             float64
AF3 Mean        float64
AF3 Std         float64
AF3 Max         float64
AF3 Min         float64
                 ...   
AF4 Mean        float64
AF4 Std         float64
AF4 Max         float64
AF4 Min         float64
eyeDetection      int64
Length: 71, dtype: object
AF3             float64
AF3 Mean        float64
AF3 Std         float64
AF3 Max         float64
AF3 Min         float64
                 ...   
AF4 Mean        float64
AF4 Std         float64
AF4 Max         float64
AF4 Min         float64
eyeDetection      int64
Length: 71, dtype: object


In [81]:
n_samples, n_inputs = train_dataset.__len__(), 70
n_outputs = 2
n_hiddens = [5, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs).to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 101
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=70, out_features=5, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=5, out_features=5, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 93.4%, Avg loss: 0.298307 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 6.6%, Avg loss: 0.778059 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 6.6%, Avg loss: 0.778044 

Done:  254.56435704231262  seconds
Test Error: 
 Accuracy: 6.6%, Avg loss: 0.778044 



In [82]:
test(test_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 6.5%, Avg loss: 0.773588 



This attempt went poorly. I quickly found that my model did not produce very good results. Imagine this was likely because the data in each particular sample was so similar. I decided to abandon this approach.

## Lumping the Data Into Chunks of Time ##


I decided I would try lumping the data into chunks of time. I thought it might be useful to lump the data in this manner so that the machine learning model could get input that was collected over a greater period of time and would hopefully be able to use this to make better predictions. In order to make this work I first needed to appropriately lump the data. I would need to ensure that each chunk only contained data for eyes-open or eyes-closed states. I would throw out ambiguous chunks. 

In [108]:
# root_directory = "D:\College\CS_545\Final_Project"
# fp_csv_file = "EEG_Eye_State_No_Outliers.csv"
# data = pd.read_csv(
#     os.path.join(root_directory, fp_csv_file), delimiter=",", usecols=range(15)
# ).dropna(axis=0)

# num_samples_in_chunk = 10
# col_labels = []
# print(data.shape)
# for i in range(num_samples_in_chunk):

#     for label in data.columns:
#         if label == "eyeDetection":
#             continue
#         col_labels.append(label + "_" + str(i))

# col_labels.append("eyeDetection")
# print(col_labels)

# new_data = []

# temp_list = []
# count = 0
# has_zero = False
# has_one = False
# zeros = 0
# ones = 0
# deleted = 0
# for index, row in data.iterrows():

#     if count == num_samples_in_chunk:
#         if has_zero and has_one:
#             count = 0
#             temp_list = []
#             print("Deleted Ambiguous Chunk")
#             deleted += 1
#         elif has_zero:
#             temp_list.append(0)
#             new_data.append(temp_list)
#             count = 0
#             temp_list = []
#             zeros += 1
#         else:
#             temp_list.append(1)
#             new_data.append(temp_list)
#             count = 0
#             temp_list = []
#             ones += 1
#         has_zero = False
#         has_one = False

#     for i in range(len(row)):

#         if i < len(row) - 1:
#             temp_list.append(row[i])
#         else:
#             if row[i] == 0:
#                 has_zero = True
#             else:
#                 has_one = True
#     count += 1
# print(len(new_data))
# print(zeros)
# print(ones)
# print("Deleted: " + str(deleted))

# new_data_df = pd.DataFrame(new_data, columns=col_labels)
# new_data_df = new_data_df.round(2)
# new_data_df = new_data_df.astype({"eyeDetection": int})
# new_data_df.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + ".csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

(14304, 15)
['AF3_0', 'F7_0', 'F3_0', 'FC5_0', 'T7_0', 'P7_0', 'O1_0', 'O2_0', 'P8_0', 'T8_0', 'FC6_0', 'F4_0', 'F8_0', 'AF4_0', 'AF3_1', 'F7_1', 'F3_1', 'FC5_1', 'T7_1', 'P7_1', 'O1_1', 'O2_1', 'P8_1', 'T8_1', 'FC6_1', 'F4_1', 'F8_1', 'AF4_1', 'AF3_2', 'F7_2', 'F3_2', 'FC5_2', 'T7_2', 'P7_2', 'O1_2', 'O2_2', 'P8_2', 'T8_2', 'FC6_2', 'F4_2', 'F8_2', 'AF4_2', 'AF3_3', 'F7_3', 'F3_3', 'FC5_3', 'T7_3', 'P7_3', 'O1_3', 'O2_3', 'P8_3', 'T8_3', 'FC6_3', 'F4_3', 'F8_3', 'AF4_3', 'AF3_4', 'F7_4', 'F3_4', 'FC5_4', 'T7_4', 'P7_4', 'O1_4', 'O2_4', 'P8_4', 'T8_4', 'FC6_4', 'F4_4', 'F8_4', 'AF4_4', 'AF3_5', 'F7_5', 'F3_5', 'FC5_5', 'T7_5', 'P7_5', 'O1_5', 'O2_5', 'P8_5', 'T8_5', 'FC6_5', 'F4_5', 'F8_5', 'AF4_5', 'AF3_6', 'F7_6', 'F3_6', 'FC5_6', 'T7_6', 'P7_6', 'O1_6', 'O2_6', 'P8_6', 'T8_6', 'FC6_6', 'F4_6', 'F8_6', 'AF4_6', 'AF3_7', 'F7_7', 'F3_7', 'FC5_7', 'T7_7', 'P7_7', 'O1_7', 'O2_7', 'P8_7', 'T8_7', 'FC6_7', 'F4_7', 'F8_7', 'AF4_7', 'AF3_8', 'F7_8', 'F3_8', 'FC5_8', 'T7_8', 'P7_8', 'O1_8', '

In [22]:
num_samples_in_chunk = 10
root_directory = "D:\College\CS_545\Final_Project"
fp_csv_file = "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + ".csv"
train_fraction = 0.8
validate_fraction = 0.1
X_pandas = pd.read_csv(
    os.path.join(root_directory, fp_csv_file),
    delimiter=",",
).dropna(axis=0)
X = X_pandas.to_numpy()

n_samples = X.shape[0]
rows = np.arange(n_samples)
np.random.shuffle(rows)

n_train = round(n_samples * train_fraction)
n_validate = round(n_samples * validate_fraction)

Xtrain = X[rows[:n_train], :]
Xvalidate = X[rows[n_train : n_train + n_validate], :]
Xtest = X[rows[n_train + n_validate :], :]
Xtrain = pd.DataFrame(Xtrain, columns=X_pandas.columns)
Xvalidate = pd.DataFrame(Xvalidate, columns=X_pandas.columns)
Xtest = pd.DataFrame(Xtest, columns=X_pandas.columns)
Xtrain = Xtrain.astype({"eyeDetection": int})
Xvalidate = Xvalidate.astype({"eyeDetection": int})
Xtest = Xtest.astype({"eyeDetection": int})
print(Xtrain.iloc[0])
print(type(X))
Xtrain.to_csv(
    "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Train.csv",
    encoding="utf-8",
    index_label=False,
    index=False,
    quoting=csv.QUOTE_NONNUMERIC,
)
Xvalidate.to_csv(
    "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Validate.csv",
    encoding="utf-8",
    index_label=False,
    index=False,
    quoting=csv.QUOTE_NONNUMERIC,
)
Xtest.to_csv(
    "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Test.csv",
    encoding="utf-8",
    index_label=False,
    index=False,
    quoting=csv.QUOTE_NONNUMERIC,
)

AF3_0           4296.92
F7_0            4012.31
F3_0            4264.62
FC5_0           4128.21
T7_0            4336.92
                 ...   
FC6_9           4222.56
F4_9            4291.28
F8_9            4631.79
AF4_9           4369.74
eyeDetection       1.00
Name: 0, Length: 141, dtype: float64
<class 'numpy.ndarray'>


In [18]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"

train_file = "EEG_Eye_State_Lumped_10_Train.csv"
validate_file = "EEG_Eye_State_Lumped_10_Validate.csv"
test_file = "EEG_Eye_State_Lumped_10_Test.csv"

# train_file = "Dummy_Train.csv"
# validate_file = "Dummy_Validate.csv"
# test_file = "Dummy_Test.csv"
count_ones_and_zeros(train_file)
count_ones_and_zeros(validate_file)
count_ones_and_zeros(test_file)
train_dataset = fpnn.FinalProjectEEGDataset(train_file, root_directory)
validate_dataset = fpnn.FinalProjectEEGDataset(validate_file, root_directory)
test_dataset = fpnn.FinalProjectEEGDataset(test_file, root_directory)
train_dataloader = DataLoader(train_dataset, batch_size=64)
validate_dataloader = DataLoader(validate_dataset, batch_size=64)
test_dataloader = DataLoader(test_dataset, batch_size=64)

Zeros:  612
Ones:  518
Zeros:  88
Ones:  54
Zeros:  76
Ones:  66
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object


In [26]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [5, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "sigmoid").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1001
train_for_epochs(epochs, train_dataloader, train_dataloader, model, loss_fn, optimizer)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=5, bias=True)
    (Sigmoid0): Sigmoid()
    (Linear1): Linear(in_features=5, out_features=5, bias=True)
    (Sigmoid1): Sigmoid()
    (Linear2): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 54.2%, Avg loss: 0.691872 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 54.2%, Avg loss: 0.689113 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 54.2%, Avg loss: 0.689113 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 54.2%, Avg loss: 0.689113 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 54.2%, Avg loss: 0.689113 

Epoch 250
-------------------------------
Test Error: 
 Accuracy: 54.2%, Avg loss: 0.689113 

Epoch 300
-------------------------------
Test Error: 
 Accuracy: 54.2%, Avg loss: 0.689113 

Ep

In [27]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667735 



In [28]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "sigmoid").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.05

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1001
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (Sigmoid0): Sigmoid()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (Sigmoid1): Sigmoid()
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (Sigmoid2): Sigmoid()
    (Linear3): Linear(in_features=10, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.694391 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.665792 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.665787 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.665785 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.665785 

Epoch 250
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.665785 



In [29]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 62.4%, Avg loss: 0.665785 



In [30]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "relu").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.1

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1001
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (ReLU2): ReLU()
    (Linear3): Linear(in_features=10, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.678976 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667083 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667083 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667083 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667083 

Epoch 250
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667083 

Epoch 300
--------

In [31]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667083 



In [32]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "relu").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.2

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1001
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (ReLU2): ReLU()
    (Linear3): Linear(in_features=10, out_features=5, bias=True)
    (ReLU3): ReLU()
    (Linear4): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.690325 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.666413 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.666431 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.666440 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.666445 

Epoch 250
------------------

In [33]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 62.4%, Avg loss: 0.666450 



In [34]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "relu").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.5

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1001
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (ReLU2): ReLU()
    (Linear3): Linear(in_features=10, out_features=5, bias=True)
    (ReLU3): ReLU()
    (Linear4): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.664654 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.663878 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.663878 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.663878 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.663878 

Epoch 250
------------------

In [36]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "relu").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.0005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1001
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (ReLU2): ReLU()
    (Linear3): Linear(in_features=10, out_features=5, bias=True)
    (ReLU3): ReLU()
    (Linear4): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.727988 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.713202 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.702363 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.694331 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.688317 

Epoch 250
------------------

In [37]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 62.4%, Avg loss: 0.668400 



In [38]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "relu").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.0005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 10001
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (ReLU2): ReLU()
    (Linear3): Linear(in_features=10, out_features=5, bias=True)
    (ReLU3): ReLU()
    (Linear4): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.675453 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.673630 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.672261 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.671229 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.670447 

Epoch 250
------------------

KeyboardInterrupt: 

In [39]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 62.4%, Avg loss: 0.667896 



In [21]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.0005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1001
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=10, out_features=5, bias=True)
    (Softplus3): Softplus(beta=1, threshold=20)
    (Linear4): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.780831 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.751333 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.730130 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 37.6%, Avg loss: 0.714763 

Epoch 200
----------------------------

In [22]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 62.4%, Avg loss: 0.668702 



https://github.com/scarca/EEG-Eyes
https://www.researchgate.net/publication/325126184_Deep_Learning_of_EEG_Time-Frequency_Representations_for_Identifying_Eye_States

The results I achieved were disappointing so I decided to try a new approach. This time I decided I would normalize the data before training.

In [48]:
# num_samples_in_chunk = 10
# root_directory = "D:\College\CS_545\Final_Project"
# fp_csv_file = "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + ".csv"
# train_fraction = 0.8
# validate_fraction = 0.1
# X_pandas = pd.read_csv(
#     os.path.join(root_directory, fp_csv_file),
#     delimiter=",",
# ).dropna(axis=0)

# X.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

# X = X_pandas.to_numpy()
# n_samples = X.shape[0]
# rows = np.arange(n_samples)
# np.random.shuffle(rows)

# n_train = round(n_samples * train_fraction)
# n_validate = round(n_samples * validate_fraction)

# Xtrain = X[rows[:n_train], :]
# Xvalidate = X[rows[n_train : n_train + n_validate], :]
# Xtest = X[rows[n_train + n_validate :], :]
# Xtrain = pd.DataFrame(Xtrain, columns=X_pandas.columns)
# Xvalidate = pd.DataFrame(Xvalidate, columns=X_pandas.columns)
# Xtest = pd.DataFrame(Xtest, columns=X_pandas.columns)
# Xtrain = Xtrain.astype({"eyeDetection": int})
# Xvalidate = Xvalidate.astype({"eyeDetection": int})
# Xtest = Xtest.astype({"eyeDetection": int})

# new = {}
# for i in Xtrain:
#     if i != "eyeDetection":
#         new[i] = (Xtrain[i] - Xtrain[i].median()) / Xtrain[i].std()
#     else:
#         new[i] = Xtrain[i]

# Xtrain = pd.DataFrame(new)

# new = {}
# for i in Xvalidate:
#     if i != "eyeDetection":
#         new[i] = (Xvalidate[i] - Xvalidate[i].median()) / Xvalidate[i].std()
#     else:
#         new[i] = Xvalidate[i]

# Xvalidate = pd.DataFrame(new)

# new = {}
# for i in Xtest:
#     if i != "eyeDetection":
#         new[i] = (Xtest[i] - Xtest[i].median()) / Xtest[i].std()
#     else:
#         new[i] = Xtest[i]

# Xtest = pd.DataFrame(new)

# Xtrain.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized_Train.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )
# Xvalidate.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized_Validate.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )
# Xtest.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized_Test.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

In [49]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"

train_file = "EEG_Eye_State_Lumped_10_Normalized_Train.csv"
validate_file = "EEG_Eye_State_Lumped_10_Normalized_Validate.csv"
test_file = "EEG_Eye_State_Lumped_10_Normalized_Test.csv"

# train_file = "Dummy_Train.csv"
# validate_file = "Dummy_Validate.csv"
# test_file = "Dummy_Test.csv"
count_ones_and_zeros(train_file)
count_ones_and_zeros(validate_file)
count_ones_and_zeros(test_file)
train_dataset = fpnn.FinalProjectEEGDataset(train_file, root_directory)
validate_dataset = fpnn.FinalProjectEEGDataset(validate_file, root_directory)
test_dataset = fpnn.FinalProjectEEGDataset(test_file, root_directory)
train_dataloader = DataLoader(train_dataset, batch_size=64)
validate_dataloader = DataLoader(validate_dataset, batch_size=64)
test_dataloader = DataLoader(test_dataset, batch_size=64)

Zeros:  621
Ones:  509
Zeros:  70
Ones:  72
Zeros:  85
Ones:  57
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object


In [23]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [20, 15, 10, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.0005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=15, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=15, out_features=10, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=10, out_features=5, bias=True)
    (Softplus3): Softplus(beta=1, threshold=20)
    (Linear4): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 40.4%, Avg loss: 0.718604 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 40.4%, Avg loss: 0.696739 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.687572 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.683372 

Epoch 200
----------------------------

In [29]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(14))) * 2,
    int(np.sqrt(np.prod(14))),
    1000,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.0005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 10000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=6, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=6, out_features=3, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=3, out_features=1000, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=1000, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.679144 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.676659 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.676233 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.675771 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.675189 

Epoch 250
-------------------------------
Test Error: 
 

In [33]:
test(validate_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 80.1%, Avg loss: 0.375153 



In [ ]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(117))) * 2,
    int(np.sqrt(np.prod(117))),
    1000,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.0005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 10000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

The results achieved here were good but I realized I had made a mistake. I had normalized the data before I split it into train, validate, and test. I needed to lump the data and then split the data and then normalize each segment to ensure that I wasn't introducing future information into my 
validate and test sets.

In [48]:
# num_samples_in_chunk = 10
# root_directory = "D:\College\CS_545\Final_Project"
# fp_csv_file = "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + ".csv"
# train_fraction = 0.8
# validate_fraction = 0.1
# X_pandas = pd.read_csv(
#     os.path.join(root_directory, fp_csv_file),
#     delimiter=",",
# ).dropna(axis=0)

# X.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

# X = X_pandas.to_numpy()
# n_samples = X.shape[0]
# rows = np.arange(n_samples)
# np.random.shuffle(rows)

# n_train = round(n_samples * train_fraction)
# n_validate = round(n_samples * validate_fraction)

# Xtrain = X[rows[:n_train], :]
# Xvalidate = X[rows[n_train : n_train + n_validate], :]
# Xtest = X[rows[n_train + n_validate :], :]
# Xtrain = pd.DataFrame(Xtrain, columns=X_pandas.columns)
# Xvalidate = pd.DataFrame(Xvalidate, columns=X_pandas.columns)
# Xtest = pd.DataFrame(Xtest, columns=X_pandas.columns)
# Xtrain = Xtrain.astype({"eyeDetection": int})
# Xvalidate = Xvalidate.astype({"eyeDetection": int})
# Xtest = Xtest.astype({"eyeDetection": int})

# new = {}
# for i in Xtrain:
#     if i != "eyeDetection":
#         new[i] = (Xtrain[i] - Xtrain[i].median()) / Xtrain[i].std()
#     else:
#         new[i] = Xtrain[i]

# Xtrain = pd.DataFrame(new)

# new = {}
# for i in Xvalidate:
#     if i != "eyeDetection":
#         new[i] = (Xvalidate[i] - Xvalidate[i].median()) / Xvalidate[i].std()
#     else:
#         new[i] = Xvalidate[i]

# Xvalidate = pd.DataFrame(new)

# new = {}
# for i in Xtest:
#     if i != "eyeDetection":
#         new[i] = (Xtest[i] - Xtest[i].median()) / Xtest[i].std()
#     else:
#         new[i] = Xtest[i]

# Xtest = pd.DataFrame(new)

# Xtrain.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized_Train.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )
# Xvalidate.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized_Validate.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )
# Xtest.to_csv(
#     "EEG_Eye_State_Lumped_" + str(num_samples_in_chunk) + "_Normalized_Test.csv",
#     encoding="utf-8",
#     index_label=False,
#     index=False,
#     quoting=csv.QUOTE_NONNUMERIC,
# )

In [17]:
root_directory = "D:\College\CS_545\Final_Project"
# root_directory = "/s/chopin/l/grad/acf003/CS_545"

train_file = "EEG_Eye_State_Lumped_10_Normalized_Train.csv"
validate_file = "EEG_Eye_State_Lumped_10_Normalized_Validate.csv"
test_file = "EEG_Eye_State_Lumped_10_Normalized_Test.csv"

# train_file = "Dummy_Train.csv"
# validate_file = "Dummy_Validate.csv"
# test_file = "Dummy_Test.csv"
count_ones_and_zeros(train_file)
count_ones_and_zeros(validate_file)
count_ones_and_zeros(test_file)
train_dataset = fpnn.FinalProjectEEGDataset(train_file, root_directory)
validate_dataset = fpnn.FinalProjectEEGDataset(validate_file, root_directory)
test_dataset = fpnn.FinalProjectEEGDataset(test_file, root_directory)
train_dataloader = DataLoader(train_dataset, batch_size=64)
validate_dataloader = DataLoader(validate_dataset, batch_size=64)
test_dataloader = DataLoader(test_dataset, batch_size=64)

Zeros:  621
Ones:  509
Zeros:  70
Ones:  72
Zeros:  85
Ones:  57
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object
AF3_0           float64
F7_0            float64
F3_0            float64
FC5_0           float64
T7_0            float64
                 ...   
FC6_9           float64
F4_9            float64
F8_9            float64
AF4_9           float64
eyeDetection      int64
Length: 141, dtype: object


In [52]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(14))) * 2,
    int(np.sqrt(np.prod(14))),
    1000,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.0005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=6, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=6, out_features=3, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=3, out_features=1000, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=1000, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 49.6%, Avg loss: 0.701164 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 51.1%, Avg loss: 0.698166 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 56.0%, Avg loss: 0.690155 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 57.4%, Avg loss: 0.677860 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 58.2%, Avg loss: 0.662290 

Epoch 250
-------------------------------
Test Error: 
 

KeyboardInterrupt: 

In [53]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(117))) * 2,
    int(np.sqrt(np.prod(117))),
    1000,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=10, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=10, out_features=1000, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=1000, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 49.6%, Avg loss: 0.693468 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 62.4%, Avg loss: 0.640390 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 70.9%, Avg loss: 0.536710 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 78.0%, Avg loss: 0.405395 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 83.0%, Avg loss: 0.319825 

Epoch 250
-------------------------------
Test Error

In [54]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(117))) * 2,
    int(np.sqrt(np.prod(117))),
    1000,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "relu").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.005

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=10, bias=True)
    (ReLU1): ReLU()
    (Linear2): Linear(in_features=10, out_features=1000, bias=True)
    (ReLU2): ReLU()
    (Linear3): Linear(in_features=1000, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 49.6%, Avg loss: 0.701064 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 70.2%, Avg loss: 0.573757 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 79.4%, Avg loss: 0.394340 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 80.9%, Avg loss: 0.316205 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 83.7%, Avg loss: 0.304453 

Epoch 250
-------------------------------
Test Error: 
 Accuracy: 83.0%, Avg loss: 0.328360 

Epoch 300
----

In [55]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(117))) * 2,
    int(np.sqrt(np.prod(117))),
    1000,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.005

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=10, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=10, out_features=1000, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=1000, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 49.6%, Avg loss: 1.033051 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 89.4%, Avg loss: 0.376740 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.583964 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.650910 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.698233 

Epoch 250
-------------------------------
Test Error

In [56]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(117))) * 2,
    int(np.sqrt(np.prod(117))),
    10,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.005

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=10, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=10, out_features=10, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=10, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 58.2%, Avg loss: 0.647764 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 85.8%, Avg loss: 0.251493 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 90.8%, Avg loss: 0.476307 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 90.1%, Avg loss: 0.537607 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 90.1%, Avg loss: 0.577488 

Epoch 250
-------------------------------
Test Error: 
 

In [57]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(117))) * 2,
    int(np.sqrt(np.prod(117))),
    100,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.005

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=10, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=10, out_features=100, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=100, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 61.7%, Avg loss: 0.640629 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.471098 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.614597 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.700515 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.765607 

Epoch 250
-------------------------------
Test Error: 

In [58]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [int(np.sqrt(np.prod(117))) * 2, int(np.sqrt(np.prod(117))), 1000, 10, 5]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.005

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=10, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=10, out_features=1000, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=1000, out_features=10, bias=True)
    (Softplus3): Softplus(beta=1, threshold=20)
    (Linear4): Linear(in_features=10, out_features=5, bias=True)
    (Softplus4): Softplus(beta=1, threshold=20)
    (Linear5): Linear(in_features=5, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 49.6%, Avg loss: 0.706832 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.355756 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 85.8%, Avg loss: 0.537157 

Epoch 150
----

I achieved pretty good results using this method. I got as high as 90.1% on validate using a fairly small number of units in one particular run. I decided to use this run as my final run against my test data. 

In [18]:
n_samples, n_inputs = train_dataset.__len__(), 140
n_outputs = 2
n_hiddens = [
    int(np.sqrt(np.prod(117))) * 2,
    int(np.sqrt(np.prod(117))),
    10,
]
model = fpnn.NeuralNetwork(n_inputs, n_hiddens, n_outputs, "softplus").to(device)
print(model)

# learning_rate = 0.05 / (n_samples * n_outputs)
learning_rate = 0.005

optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
loss_fn = torch.nn.CrossEntropyLoss()
epochs = 1000
train_for_epochs(
    epochs, train_dataloader, validate_dataloader, model, loss_fn, optimizer
)

NeuralNetwork(
  (flatten): Flatten(start_dim=1, end_dim=-1)
  (linear_stack): Sequential(
    (Linear0): Linear(in_features=140, out_features=20, bias=True)
    (ReLU0): ReLU()
    (Linear1): Linear(in_features=20, out_features=10, bias=True)
    (Softplus1): Softplus(beta=1, threshold=20)
    (Linear2): Linear(in_features=10, out_features=10, bias=True)
    (Softplus2): Softplus(beta=1, threshold=20)
    (Linear3): Linear(in_features=10, out_features=2, bias=True)
  )
)
Epoch 0
-------------------------------
Test Error: 
 Accuracy: 59.6%, Avg loss: 0.658107 

Epoch 50
-------------------------------
Test Error: 
 Accuracy: 85.8%, Avg loss: 0.562668 

Epoch 100
-------------------------------
Test Error: 
 Accuracy: 88.7%, Avg loss: 0.712021 

Epoch 150
-------------------------------
Test Error: 
 Accuracy: 87.9%, Avg loss: 0.792777 

Epoch 200
-------------------------------
Test Error: 
 Accuracy: 87.9%, Avg loss: 0.853813 

Epoch 250
-------------------------------
Test Error: 
 

In [19]:
test(test_dataloader, model, loss_fn)

Test Error: 
 Accuracy: 90.1%, Avg loss: 1.461865 



In [35]:
nb_classes = 2

confusion_matrix = torch.zeros(nb_classes, nb_classes)
with torch.no_grad():
    for i, (inputs, classes) in enumerate(test_dataloader):
        inputs = inputs.to(device)
        classes = classes.to(device)
        outputs = model(inputs.float())
        _, preds = torch.max(outputs, 1)
        for t, p in zip(classes.view(-1), preds.view(-1)):
            confusion_matrix[t.long(), p.long()] += 1

print(confusion_matrix)
new_confusion_matrix = torch.zeros(nb_classes, nb_classes)
for i, row in enumerate(confusion_matrix):
    sum = row.sum()
    new_confusion_matrix[i, 0] = 100 * row[0] / sum
    new_confusion_matrix[i, 1] = 100 * row[1] / sum
conf_matrix = pd.DataFrame(
    new_confusion_matrix.numpy(), index=["open", "closed"], columns=["open", "closed"]
)
# cf.style.background_gradient(cmap='Blues').format("{:.1f} %")
print("Percent Correct")
conf_matrix.style.background_gradient(cmap="Blues").format("{:.1f}")
# print(confusion_matrix)

tensor([[75., 10.],
        [ 4., 52.]])
Percent Correct


The results I achieved were pretty good. Overall on the test data I had 90.1% accuracy.When I broke it out as open versus closed my model was correctly identifying cases of open eyes 88.2% of the time and closed eyes 92.9% of the time. This was a fairly even split and showed that the model was not using a strategy biased in favor of one State versus the other. I was pretty satisfied with these results at classifying using a standard classification neural network. 